hep/proposals/0000-decorators.md

```diff
@@ -0,0 +1,299 @@
+
+# Motivation
+[motivation]: #motivation
+
+The current context manager based syntax is good but it doesn't map well to template types that contain inputs/outputs. With the `@script` decorator, we have made hera feel extremely Pythonic and it has the added benefit of being runnable locally. We wish to extend this syntax to other template types which will allow for better readability and local testing and it will also standardize the template interface for Argo Workflows.
+
+# Proposal
+
+This HEP proposes that every template be mapped into a decorated function with HeraIO classes as the inputs and outputs.
+
+# Code Examples
+
+## Basic Workflow
+
+```python
+from hera.workflows import WorkflowTemplate
+from hera import io as hio
+
+# We start by defining our Workflow Template
+wt = WorkflowTemplate(name="my-template")
+
+# This defines the a template input
+class MyInput(hio.Input):
+    user: str
+
+@wt.entrypoint  # Sets hello_world as the default entrypoint for the workflow template
+@wt.script  # Adds a new script template to the workflow template called hello_world
+def hello_world(my_input: MyInput) -> hio.Output:
+    output = hio.Output()
+    output.result = f"Hello Hera User: {my_input.user}!"
+    return output
+
+# Allows users to instantiate a workflow template into a workflow
+workflow = wt.run(MyInput(name="happy-hera-user"), wait=True)
+# Returned workflow is a normal hera Workflow class and can be accessed as normal
+assert workflow.status.phase == "Succeeded"
+assert workflow.status.outputs.result == "Hello Hera User: happy-hera-user!"
+```
+
+## DAG Template
+
+```python
+from hera.workflows import WorkflowTemplate
+from hera import io as hio
+from hera.expr import spring
+
+# We start by defining our Workflow Template
+wt = WorkflowTemplate(name="my-template")
+
+@wt.script  # Adds a new script template to the workflow template
+def setup() -> hio.Output:
+    return hio.Output(result="Setting things up")
+
+class ConcatInput(hio.Input):
+    word_a: str
+    word_b: str
+
+@wt.script  # Adds a new script template to the workflow template
+def concat(concat_input: ConcatInput) -> hio.Output:
+    return hio.Output(result=f"{concat_input.word_a} {concat_input.word_b}")
+
+class WorkerInput(hio.Input):
+    value_a: str
+    value_b: str
+
+class WorkerOutput(hio.Output):
+    value: str
+
+
+# The great outcome of this new syntax is that the template definition
+# vs the template call becomes very explicit.
+# When a template is defined, it is defined via the decorator which registers it
+# When a template is invoked, it is invoked either via the callable syntax or via
+# wt.run() if it is the entrypoint
+# Another great side-effect is that all the templates can also be "run" locally as functions
+@wt.entrypoint
+@wt.dag
+def worker(worker_input: WorkerInput) -> WorkerOutput:
+    # We will need to use something like python-varname
+    # to automatically create a task called "setup_task" here.
+    setup_task = setup()
```

Comment:
Would be worth pointing out `setup_task` will be an `hio.Output` when running locally/testing, but then a `Task` when building the workflow - and that we can only realistically do this by limiting inputs/outputs to the Hera IO classes